### PR TITLE
Fold messages by uid

### PR DESCRIPTION
**Pull request by IsaackMwenda**
Review https://github.com/AfricasVoices/Test-Pipeline-Engagement-DB/pull/26 first

### CODE REVIEW

File: src/engagement_db_to_analysis/engagement_db_to_analysis.py [hunk 1]
```diff
@@ -117,9 +117,54 @@ def _convert_messages_to_traced_data(user, messages_map):
...
+
+        participant_uuid = message_td["participant_uuid"]
+        update_key = message_td["dataset"]
+
+        # Check if we already have the participant traced data in individuals
+        if message_td["participant_uuid"] in individuals_uids:
+
+            # if it exists
+            # update the existing individual traced data with the message_td
+            for ind_td in individuals_td:
```
Comment (as2388) [inline src/engagement_db_to_analysis/engagement_db_to_analysis.py:144]: If individuals_td was a dict of uuid -> individual then you don't need to maintain a separate individuals_uids set or pay to loop over this list for every message

File: src/engagement_db_to_analysis/engagement_db_to_analysis.py [hunk 2]
```diff
@@ -117,9 +117,54 @@ def _convert_messages_to_traced_data(user, messages_map):
...
+
+    individuals_td = []
+    individuals_uids = set()  # for uid lookup
+    for message_td in message_traced_data:
+
+        participant_uuid = message_td["participant_uuid"]
+        update_key = message_td["dataset"]
+
+        # Check if we already have the participant traced data in individuals
+        if message_td["participant_uuid"] in individuals_uids:
```
Comment (as2388) [inline src/engagement_db_to_analysis/engagement_db_to_analysis.py:140]: If you restructure this to follow the pattern:
if ind not in collection:
   create ind
add message to ind

then you don't need to maintain so many branches of logic.

File: src/engagement_db_to_analysis/engagement_db_to_analysis.py [hunk 3]
```diff
@@ -117,9 +117,54 @@ def _convert_messages_to_traced_data(user, messages_map):
...
+        else:
+
+            # if it does not exists
+            # 1. create an individual traced data for this uid
+            ind_dict = {"participant_uuid": participant_uuid}
+            ind_td = TracedData(ind_dict, Metadata(user,
+                                                   Metadata.get_call_location(), TimeUtils.utc_now_as_iso_string()))
+
+            # 2. update the individual traced data with the message_td
+            ind_td.append_data({update_key: [message_td.serialize()]},
```
Comment (as2388) [inline src/engagement_db_to_analysis/engagement_db_to_analysis.py:161]: What if they already have a message at this update_key?

File: src/engagement_db_to_analysis/engagement_db_to_analysis.py [hunk 4]
```diff
@@ -97,7 +97,7 @@ def _convert_messages_to_traced_data(user, messages_map):
     """
     Converts messages dict objects to TracedData objects.
 
-    :param user:
+    :param user: identifier of user running the pipeline.
```
[the remark this answers is not in the document]
Comment (IsaackMwenda) [reply src/engagement_db_to_analysis/engagement_db_to_analysis.py]: updated thanks.

File: src/engagement_db_to_analysis/engagement_db_to_analysis.py [hunk 5]
```diff
@@ -117,9 +117,54 @@ def _convert_messages_to_traced_data(user, messages_map):
...
+    :type user: str
+    :param message_traced_data: Messages TracedData objects to group.
+    :type message_traced_data: list of TracedData
+    :return: individual TracedData objects.
+    :rtype: list of individual TracedData objects.
+    """
+
+    individuals_td = []
+    individuals_uids = set()  # for uid lookup
+    for message_td in message_traced_data:
```
[the remark this answers is not in the document]
Comment (IsaackMwenda) [reply src/engagement_db_to_analysis/engagement_db_to_analysis.py:134]: Renamed iterable to participants_traced_data and single to participant where aplicable.

File: src/engagement_db_to_analysis/engagement_db_to_analysis.py [hunk 6]
```diff
@@ -117,9 +117,54 @@ def _convert_messages_to_traced_data(user, messages_map):
...
+
+        participant_uuid = message_td["participant_uuid"]
+        update_key = message_td["dataset"]
```
[the remark this answers is not in the document]
Comment (IsaackMwenda) [reply src/engagement_db_to_analysis/engagement_db_to_analysis.py:137]: Renamed to message_dataset

File: src/engagement_db_to_analysis/engagement_db_to_analysis.py [hunk 7]
```diff
@@ -117,9 +117,54 @@ def _convert_messages_to_traced_data(user, messages_map):
...
+
+    :param user: identifier of user running the pipeline.
+    :type user: str
+    :param message_traced_data: Messages TracedData objects to group.
+    :type message_traced_data: list of TracedData
+    :return: individual TracedData objects.
+    :rtype: list of individual TracedData objects.
+    """
+
+    individuals_td = []
```
[the remark this answers is not in the document]
Comment (IsaackMwenda) [reply src/engagement_db_to_analysis/engagement_db_to_analysis.py]: updated thanks.

File: src/engagement_db_to_analysis/engagement_db_to_analysis.py (hunk 1 above)
Comment (IsaackMwenda) [reply src/engagement_db_to_analysis/engagement_db_to_analysis.py]: Thanks wanted to retain individuals as a list of traced data but I think there we can have it as a dict too.
Refactored  to a dictionary of uid -> td objects i.e
` participants_traced_data = {'avf-participant-uuid': {s01e01: [msg1, msg2], age: [msg1]}`

File: src/engagement_db_to_analysis/engagement_db_to_analysis.py (hunk 2 above)
Comment (IsaackMwenda) [reply src/engagement_db_to_analysis/engagement_db_to_analysis.py]: Updated - thanks for the suggestions.

File: src/engagement_db_to_analysis/engagement_db_to_analysis.py (hunk 3 above)
Comment (IsaackMwenda) [reply src/engagement_db_to_analysis/engagement_db_to_analysis.py]: We append the new message to the key,

File: src/engagement_db_to_analysis/engagement_db_to_analysis.py [hunk 8]
```diff
@@ -97,37 +97,72 @@ def _convert_messages_to_traced_data(user, messages_map):
...
+    participants_traced_data = {}
+    for message in messages_traced_data:
+
+        participant_uuid = message["participant_uuid"]
+        message_dataset = message["dataset"]
+
+        if message["participant_uuid"] not in participants_traced_data.keys():
+
+            participant_td = TracedData({message_dataset: [message.serialize()]}, Metadata(user,
+                                                                                           Metadata.get_call_location(),
```
Comment (as2388) [inline src/engagement_db_to_analysis/engagement_db_to_analysis.py:141]:
```suggestion
    """
    participants_traced_data = {}
    for message in messages_traced_data:
        participant_uuid = message["participant_uuid"]
        message_dataset = message["dataset"]

        if message["participant_uuid"] not in participants_traced_data.keys():
            participant_td = TracedData({message_dataset: [message.serialize()]}, 
                             Metadata(user, Metadata.get_call_location(),
```
formatting nits

File: src/engagement_db_to_analysis/engagement_db_to_analysis.py [hunk 9]
```diff
@@ -97,37 +97,72 @@ def _convert_messages_to_traced_data(user, messages_map):
...
+
+    :param user: Identifier of user running the pipeline.
+    :type user: str
+    :param messages_traced_data: Messages TracedData objects to group.
+    :type messages_traced_data: list of TracedData
+    :return: Individual TracedData objects.
+    :rtype: dict of uid -> individual TracedData objects.
+    """
+
+    participants_traced_data = {}
```
Comment (as2388) [inline src/engagement_db_to_analysis/engagement_db_to_analysis.py:132]:
```suggestion
    participants_traced_data_map = {}
```
since it's not traced data

**Review by as2388 — APPROVED** [review]

File: src/engagement_db_to_analysis/engagement_db_to_analysis.py [hunk 10]
```diff
@@ -97,37 +97,72 @@ def _convert_messages_to_traced_data(user, messages_map):
...
+
+            participant_td = TracedData({message_dataset: [message.serialize()]}, Metadata(user,
+                                                                                           Metadata.get_call_location(),
+                                                                                           TimeUtils.utc_now_as_iso_string()))
+            participants_traced_data[participant_uuid] = participant_td
+
+        else:
+
+            if message_dataset in participants_traced_data[participant_uuid].keys():
+                participants_traced_data[participant_uuid][message_dataset].append(message)
```
[the remark this answers is not in the document]
Comment (IsaackMwenda) [reply src/engagement_db_to_analysis/engagement_db_to_analysis.py]: Updated thanks.

File: src/engagement_db_to_analysis/engagement_db_to_analysis.py (hunk 9 above)
Comment (IsaackMwenda) [reply src/engagement_db_to_analysis/engagement_db_to_analysis.py]: Updated thanks.

**Pull request merged**